churchland_pipeline_python/utilities/datajointutils.py

```diff
@@ -8,106 +8,173 @@
 
-    Args:
-        names (list): Strings of table names
-        context (frame, optional): Frame used to evaluate table names. Defaults to [].
+def insertpart(master: DataJointTable, part_name: str, **kwargs) -> None:
+    """Inserts an entry to master-part tables, given the master table, the part table name,
+    and a set of keyword arguments containing the part table attributes. Checks to ensure that
+    the keyword arguments are all valid attributes of the part table. Assumes that entries in 
+    the master table are uniquely identified by an ID number. Creates a new ID for each new entry
+    so that every entry is uniquely identifiable by its ID across all part tables."""
 
-    Returns:
-        tables (list): DataJoint tables
-    """
+    # read master table attributes
+    master_table_attr = readattributes(master)
 
-    if context:
-        tables = [eval(dj.table.lookup_class_name(x, context.f_globals), context.f_globals) for x in names]
-    else:
-        for frame in inspect.stack():
-            try:
-                eval(dj.table.lookup_class_name(names[0], frame[0].f_globals), frame[0].f_globals)
-            except TypeError:
-                pass
-            else:
-                tables = [eval(dj.table.lookup_class_name(x, frame[0].f_globals), frame[0].f_globals) for x in names]
-    
-    return tables
+    # validate master table format
+    assert len(master_table_attr) == 1, 'Master table has more than one attribute'
 
-def nextkey(table, index=0):
-    """Gets the next (unpopulated) key for a table.
+    master_attr_name = next(iter(master_table_attr))
+    assert '_id' in master_attr_name, 'Expected variable ID as master table attribute'
 
-    Args:
-        table (datajoint.user_tables.OrderedClass): DataJoint table
-        index (int, optional): Index of next key to fetch. Defaults to 0.
+    try:
+        # get part table
+        part = getattr(master, part_name)
 
-    Returns:
-        key (dict): Primary key
-    """
+        # read part table attributes
+        part_table_attr = readattributes(part)
+
+        # ensure keyword keys are members of secondary attributes list
+        assert set(kwargs.keys()).issubset(set(part_table_attr.keys())), 'Unrecognized keyword argument(s)'
+
+        # check if entry already exists in table
+        if part():
+
+            # append default values if missing
+            for key,val in part_table_attr.items():
+                if key not in kwargs.keys() and not math.isnan(val):
+                    kwargs.update({key:val})
+
+            # existing entries
+            part_entity = part.fetch(as_dict=True)
+            part_entity_attr = [{k:float(v) for k,v in entity.items() \
+                if k != master_attr_name and v is not None} \
+                for entity in part_entity]
+            
+            # cross reference
+            assert not any([kwargs == entity_attr for entity_attr in part_entity_attr]), 'Duplicate entry!'
```

Comment:
if you remove the square brackets, making this a generator expression rather than a list comprehension, then the execution becomes more efficient, quitting after the fist `True` is found.